src/array/cpu/csr_linegraph.cc

```diff
@@ -0,0 +1,60 @@
+/*!
+ *  Copyright (c) 2020 by Contributors
+ * \file array/cpu/coo_line_graph.cc
+ * \brief CSR LineGraph
+ */
+
+#include <dgl/array.h>
+#include <numeric>
+#include <algorithm>
+#include <vector>
+#include <iterator>
+
+namespace dgl {
+namespace aten {
+namespace impl {
+
+template <DLDeviceType XPU, typename IdType>
+CSRMatrix CSRLineGraph(const CSRMatrix &csr, bool backtracking) {
```

Comment:
One problem. The node order in the line graph is the same as the edge order in the original graph but this implementation does not keep that order. I think one quick implementation is just convert it to COO, create a line graph and then convert to CSR again. Since the operation is quite rare, the performance is not critical. We can also prefer COO format.

Reply:
One problem of this solution is if the format is restricted to CSR? converting to COO will case extra memory cost.

Reply:
I think we can do this. Otherwise the CSR implementation is very complex.